### PR TITLE
convert resource to trait

### PR DESCRIPTION
**Pull request by bmacer**

### CODE REVIEW

**Comment by Swader** [comment]
I don't see the resource override here, or am I missing it?

File: traits/src/resource.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	/// If the resource has the thumb property, this will be a URI to a thumbnail of the given
+	/// resource. For example, if we have a composable NFT like a Kanaria bird, the resource is
+	/// complex and too detailed to show in a search-results page or a list. Also, if a bird owns
+	/// another bird, showing the full render of one bird inside the other's inventory might be a
+	/// bit of a strain on the browser. For this reason, the thumb value can contain a URI to an
+	/// image that is lighter and faster to load but representative of this resource.
+	pub thumb: Option<BoundedString>,
+}
+
+/// Abstraction over a Collection system.
```
Comment (ilionic) [inline traits/src/resource.rs:46]: Abstraction over a **Resource** system.

File: pallets/rmrk-core/src/lib.rs [hunk 2]
```diff
@@ -490,37 +491,19 @@ pub mod pallet {
...
-			if nft.rootowner != sender.unwrap_or_default() {
-				pending = true;
-			}
-
-			let resource_id = Self::get_next_resource_id()?;
-			ensure!(
-				Resources::<T>::get((collection_id, nft_id, resource_id)).is_none(),
-				Error::<T>::ResourceAlreadyExists
-			);
+			let sender = Self::ensure_protocol_or_signed(origin)?.unwrap_or_default();
```
Comment (ilionic) [inline pallets/rmrk-core/src/lib.rs:494]: please pull `main` 
should be simply `let sender = ensure_signed(origin.clone())?;`

File: pallets/rmrk-core/src/functions.rs [hunk 3]
```diff
@@ -1,5 +1,69 @@
...
+		let empty = base.is_none()
+			&& src.is_none()
+			&& metadata.is_none()
+			&& slot.is_none()
+			&& license.is_none()
+			&& thumb.is_none();
+		ensure!(!empty, Error::<T>::EmptyResource);
+
+		let res =
+			ResourceInfo { id: resource_id, base, src, metadata, slot, license, thumb, pending };
```
Comment (ilionic) [inline pallets/rmrk-core/src/functions.rs:36]: consider simplifying this bit:
```
pending: nft.rootowner != sender,
```
so no need for extra `pending` mut variable

**Review by ilionic — APPROVED** [review]
LGTM

File: pallets/rmrk-core/src/lib.rs (hunk 2 above)
Comment (bmacer) [reply pallets/rmrk-core/src/lib.rs]: resolved in https://github.com/rmrk-team/rmrk-substrate/pull/41/commits/e8c89c2346bbfcb3104727e6bb296de4646b619f

File: traits/src/resource.rs (hunk 1 above)
Comment (bmacer) [reply traits/src/resource.rs]: https://github.com/rmrk-team/rmrk-substrate/pull/41/commits/9872569c887792996e5d513e8e6c8d450df82bdc

File: pallets/rmrk-core/src/functions.rs (hunk 3 above)
Comment (bmacer) [reply pallets/rmrk-core/src/functions.rs]: https://github.com/rmrk-team/rmrk-substrate/pull/41/commits/cc295151e72e9728c6e3e0c1f22747c3070a3c0b

**Pull request merged**